pyuvdata/uvdata/uvdata.py

```diff
@@ -2574,6 +2574,40 @@ def check(
 
         return True
 
+    def __eq__(self, other, check_extra=True, allowed_failures=None):
+        """
+        Test for equality between two objects.
+
+        By default, this will add the `filename` attribute to the list of
+        parameters allowed to be not equal while passing an overall equality
+        check. The user is permitted to add other attributes to this list with
+        the `allowed_failures` keyword argument.
```

Comment:
I just realized that this approach does not let a user require that filenames are the same because we always add that parameter. Another option would be to set the default like this: `allowed_failures=["filename"]`. The downside there being that the user would have to include `"filename"` in the list themselves, but if a user is modifying the default to this magic method they probably know what they are doing.

Reply:
Also it seems a little silly to me that someone would need the filename to be equal, but I'm sure there's a reasonable case for it. Isn't it more important all the (meta)data is the same in the end?

Reply:
I can think of a _potential_ (but not yet supported) use-case where you want the `UVData` object to have only the metadata, and load the actual visibilities on demand. Something like this is implemented in the `MirParser` class -- may be useful for dealing w/ larger file sizes in the future. But again, it's not actually a supported thing (yet...).

Reply:
I think I'm a little confused about the filename here though? Wouldn't all the metadata being equal tell you it's the same dataset?

Reply:
I suppose that's probably true -- or rather, you'd have to have something that _only_ acted on the data but not the metadata. Only thing that comes to mind would be applying flagging or gains, although I don't know if performing those actions touches the metadata in some way that I'm not aware of.

Reply:
I think I just don't understand the point you're trying to make.

Reply:
If I understand @kartographer's point correctly, I think it's similar to a change I made for partial writing in UVH5: 

https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/4c66208245db29be8b9920434d6d48d23b29e367/pyuvdata/uvdata/uvh5.py#L1600-L1677

In this case, we're checking that the partially-written file on disk matches the object in memory, but unless the user is overwriting the file (bad idea!), the `filename` won't match. I explicitly ignore that parameter when testing equality. But I think the case mentioned is sort of the inverse, where you want to make sure the filename in the partially-loaded object matches the one you ask to fetch more data from (right)? In that case, the function could maybe just check that `filename` matches explicitly.

Reply:
Sorry, I think I'm getting a little overly in the weeds on this -- let me take it back a step. I can imagine a case where one might want to know _if_ to UVData objects came from the same file versus two different files (even if the data they contain is identical). These cases that I can imagine are potentially corner-cases, but I think you want to have the user be able to expose that capability if it was so desired.

I think the core issue here isn't so much that this particular parameter (`UVData.filename`) is excluded by default, it's that `allowed_failures=None` is being treated as synonymous with some default settings, rather than having `allowed_failures=['filename']` actually being set explicitly in the function definition. I can see users wanting to have the control afforded by the latter, wherever possible.

Reply:
Okay I think I see that. We could make the default a tuple `("filename")` that would be immutable at least and get around the really awful things that happen with mutable kwargs in functions :thinking: Though that definitely be different than most of the rest of the codebase.

Reply:
Addressed in 00cb301.

Reply:
This looks good now, thank you! I'm not sure that you still need the longer explanation in the main part of the docstring since the default is made clear in the Parameters section.

Reply:
Good point, removed in 2b832011.